### PR TITLE
Add metadata support

### PR DESCRIPTION
**Pull request by acerv**
This is a first step toward parallel execution. We read metadata provided by the new LTP API and check if test can run in parallel or not. The choice is based on flags we are reading from metadata which are defined inside the ltp.data.PARALLEL_BLACKLIST list.

### CODE REVIEW

File: ltp/data.py [hunk 1]
```diff
@@ -90,28 +99,57 @@ def arguments(self):
...
     :type content: str
+    :param metadata: metadata JSON file content
+    :type metadata: dict
     :returns: Suite
     """
     if not content:
         raise ValueError("content is empty")
 
-    LOGGER.info("collecting testing suite: %s", suite_name)
+    LOGGER.info("Collecting testing suite: %s", suite_name)
```
Comment (pevik) [inline ltp/data.py:140]: nit for next time: unrelated changes are better to be in separate commit (easier to review).

**Comment by pevik** [comment]
On a very brief look LGTM. BTW that reminds me that shell API does not have (yet) metadata support.

**Comment by pevik** [comment]
```
Reviewed-by: Petr Vorel <pvorel@suse.cz>
Tested-by: Petr Vorel <pvorel@suse.cz>
```

File: ltp/data.py [hunk 2]
```diff
@@ -90,28 +99,57 @@ def arguments(self):
...
 
-def read_runtest(suite_name: str, content: str) -> Suite:
+PARALLEL_BLACKLIST = [
+    "needs_root",
+    "needs_device",
+    "mount_device",
+    "mntpoint",
+    "resource_file",
+    "format_device",
+    "forks_child"
```
Comment (richiejp) [inline ltp/data.py]: I don't think forks child is a very good indicator. A lot of tests just fork a child to test the process of forking/cloning or unshare, cgroup hierarchy etc. Possibly a better one would be if max_runtime > 5 (assuming you want to avoid running resource intensive tests in parallel).

File: ltp/data.py [hunk 3]
```diff
@@ -90,28 +99,57 @@ def arguments(self):
...
+    def parallelizable(self):
+        """
+        If True, test can be run in parallel.
+        """
+        return self._parallelizable
+
 
-def read_runtest(suite_name: str, content: str) -> Suite:
+PARALLEL_BLACKLIST = [
+    "needs_root",
```
Comment (richiejp) [inline ltp/data.py:111]: I think we also need save_restore, because sometimes this is set, but not needs_root.

Comment (acerv) [reply ltp/data.py:111]: @richiejp we discussed a few meetings ago about avoiding tests which need root privileges.

Comment (richiejp) [reply ltp/data.py:111]: yes, I'm not suggesting we remove needs_root. I am saying we also need save_restore.

File: ltp/data.py [hunk 4]
```diff
@@ -124,16 +162,51 @@ def read_runtest(suite_name: str, content: str) -> Suite:
...
 
-        test = Test(test_name, test_cmd, test_args)
+        parallelizable = True
+
+        if not metadata_tests:
+            # no metadata no party
+            parallelizable = False
+        else:
+            test_params = None
+            for mtest_name, mtest_params in metadata_tests.items():
```
Comment (richiejp) [inline ltp/data.py]: Potentially there could be 10K+ tests in the metadata and 2-3K lines in a runtest file. We probably want a dict of mtest_name -> mtest_params.

File: ltp/data.py [hunk 5]
```diff
@@ -124,16 +163,47 @@ def read_runtest(suite_name: str, content: str) -> Suite:
...
+        if not parallelizable:
+            LOGGER.info("Test '%s' is not parallelizable", test_name)
+        else:
+            LOGGER.info("Test '%s' is parallelizable", test_name)
+
+        test = Test(
+            test_name,
+            test_cmd,
+            test_args,
+            parallelizable=parallelizable)
```
Comment (metan-ucw) [inline ltp/data.py:196]: I guess that we will eventually make use of more test metadata, so maybe we should store the whole test_params into the test object, but I guess that we can do that later on.

**Comment by metan-ucw** [comment]
Looks like a good start.

Reviewed-by: Cyril Hrubis <chrubis@suse.cz>

**Comment by acerv** [comment]
This branch will be merged with parallel execution support.
